packages/core/src/modules/dcql/DcqlService.ts

```diff
@@ -0,0 +1,277 @@
+
+    const mdocApi = this.getMdocApi(agentContext)
+    const mdocRecords = formats.has('mso_mdoc') ? await mdocApi.getAll() : []
+    allRecords.push(...mdocRecords)
+
+    return allRecords
+  }
+
+  public async getCredentialsForRequest(agentContext: AgentContext, dcqlQuery: DcqlQuery): Promise<DcqlQueryResult> {
+    const credentialRecords = await this.queryCredentialForPresentationDefinition(agentContext, dcqlQuery)
+
+    const mappedCredentials: DcqlCredentialRepresentation[] = credentialRecords.map((record) => {
+      if (record.type === 'MdocRecord') {
+        return {
+          docType: record.getTags().docType,
+          namespaces: Mdoc.fromBase64Url(record.base64Url).issuerSignedNamespaces,
+        } satisfies DcqlMdocRepresentation
+      } else if (record.type === 'SdJwtVcRecord') {
+        return {
+          vct: record.getTags().vct,
+          claims: this.getSdJwtVcApi(agentContext).fromCompact(record.compactSdJwtVc)
+            .prettyClaims as DcqlSdJwtVcRepresentation.Claims,
+        } satisfies DcqlSdJwtVcRepresentation
+      } else {
+        // TODO:
+        throw new DcqlError('W3C credentials are not supported yet')
+      }
+    })
+
+    const queryResult = DcqlQuery.query(dcqlQuery, mappedCredentials)
+    const matchesWithRecord = Object.fromEntries(
+      Object.entries(queryResult.credential_matches).map(([credential_query_id, result]) => {
+        return [credential_query_id, { ...result, record: credentialRecords[result.credential_index] }]
+      })
+    )
+
+    return {
+      ...queryResult,
+      credential_matches: matchesWithRecord,
+    }
+  }
+
+  /**
+   * Selects the credentials to use based on the output from `getCredentialsForRequest`
+   * Use this method if you don't want to manually select the credentials yourself.
+   */
+  public selectCredentialsForRequest(dcqlQueryResult: DcqlQueryResult): DcqlCredentialsForRequest {
+    if (!dcqlQueryResult.canBeSatisfied) {
+      throw new DcqlError(
+        'Cannot select the credentials for the dcql query presentation if the request cannot be satisfied'
+      )
+    }
+
+    const credentials: DcqlCredentialsForRequest = {}
+
+    if (dcqlQueryResult.credential_sets) {
+      for (const credentialSet of dcqlQueryResult.credential_sets) {
+        // undefined defaults to true
+        if (credentialSet.required === false) continue
+        const firstFullFillableOption = credentialSet.options.find((option) =>
+          option.every((credential_id) => dcqlQueryResult.credential_matches[credential_id].success)
+        )
+
+        if (!firstFullFillableOption) {
+          throw new DcqlError('Invalid dcql query result. No option is fullfillable')
+        }
+
+        for (const credentialQueryId of firstFullFillableOption) {
+          const credential = dcqlQueryResult.credential_matches[credentialQueryId]
+
+          if (credential.success && credential.record.type === 'MdocRecord' && 'namespaces' in credential.output) {
```

Comment:
I usually prefer adding a diferentiator key that then also defiens the type of credential.

So credential has a format key and if it is mso_mdoc then output has namespaces.